streams/src/test/java/org/apache/kafka/streams/integration/AbstractJoinIntegrationTest.java

```diff
@@ -87,7 +87,37 @@ public static Collection<Object[]> data() {
 
     StreamsBuilder builder;
 
-    private final List<Input<String>> input = Arrays.asList(
+    protected final List<Input<String>> input = Arrays.asList(
+        new Input<>(INPUT_TOPIC_LEFT, null, 1),
+        new Input<>(INPUT_TOPIC_RIGHT, null, 2),
+        new Input<>(INPUT_TOPIC_LEFT, "A", 3),
+        new Input<>(INPUT_TOPIC_RIGHT, "a", 4),
+        new Input<>(INPUT_TOPIC_LEFT, "B", 5),
+        new Input<>(INPUT_TOPIC_RIGHT, "b", 6),
+        new Input<>(INPUT_TOPIC_LEFT, null, 7),
+        new Input<>(INPUT_TOPIC_RIGHT, null, 8),
+        new Input<>(INPUT_TOPIC_LEFT, "C", 9),
+        new Input<>(INPUT_TOPIC_RIGHT, "c", 10),
+        new Input<>(INPUT_TOPIC_RIGHT, null, 11),
+        new Input<>(INPUT_TOPIC_LEFT, null, 12),
+        new Input<>(INPUT_TOPIC_RIGHT, null, 13),
+        new Input<>(INPUT_TOPIC_RIGHT, "d", 7), // out-of-order data with null as latest
+        new Input<>(INPUT_TOPIC_LEFT, "D", 6),
+        new Input<>(INPUT_TOPIC_LEFT, null, 2),
+        new Input<>(INPUT_TOPIC_RIGHT, null, 3),
+        new Input<>(INPUT_TOPIC_RIGHT, "e", 14),
+        new Input<>(INPUT_TOPIC_LEFT, "E", 15),
+        new Input<>(INPUT_TOPIC_LEFT, null, 10), // out-of-order data with non-null as latest
+        new Input<>(INPUT_TOPIC_RIGHT, null, 9),
+        new Input<>(INPUT_TOPIC_LEFT, "F", 4),
+        new Input<>(INPUT_TOPIC_RIGHT, "f", 3)
+    );
+
+    // used for stream-stream join tests where out-of-order data does not meaningfully affect
+    // the result, and the main `input` list results in too many result records/test noise.
+    // also used for table-table multi-join tests, since out-of-order data with table-table
+    // joins is already tested in non-multi-join settings.
+    protected final List<Input<String>> inputWithoutOutOfOrderData = Arrays.asList(
```

Comment:
As suggested in https://github.com/apache/kafka/pull/13497#discussion_r1163397015, I have split the input test data into two copies: a smaller one without out-of-order data (for use in stream-stream join and table-table multi-join tests) and a larger one with out-of-order data (for use elsewhere, including to validate versioned joins).